iidm/iidm-api/src/main/java/com/powsybl/iidm/network/limitreduction/computer/AbstractReducedLimitsComputer.java

```diff
@@ -0,0 +1,89 @@
+ */
+package com.powsybl.iidm.network.limitreduction.computer;
+
+import com.powsybl.iidm.network.LimitType;
+import com.powsybl.iidm.network.ThreeSides;
+import com.powsybl.iidm.network.limitreduction.result.LimitsContainer;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * <p>Abstract class responsible for computing reduced limits.</p>
+ * <p>Already computed reduced limits are stored in an internal cache to avoid unnecessary computations.
+ * This cache should be cleared with {@link #clearCache()} when the reductions to apply are changed.</p>
+ *
+ * @author Olivier Perrin {@literal <olivier.perrin at rte-france.com>}
+ */
+public abstract class AbstractReducedLimitsComputer<P, L> implements ReducedLimitsComputer<P, L> {
+    private final Map<CacheKey<P>, LimitsContainer<L>> reducedLimitsCache;
+
+    protected AbstractReducedLimitsComputer() {
+        this.reducedLimitsCache = new HashMap<>();
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>Resulting limits are stored in an internal cache, by (identifier's ID, limit type, side), in order to avoid
+     * unnecessary re-computation. If needed, this cache can be cleared using {@link #clearCache()}.</p>
+     *
+     * @param processable The network element for which the reduced limits must be computed
+     * @param limitType The type of the limits to process
+     * @param side The side of <code>processable</code> on which the limits should be retrieved
+     * @return an object containing the original limits and the altered ones
+     */
+    @Override
+    public Optional<LimitsContainer<L>> getLimitsWithAppliedReduction(P processable, LimitType limitType, ThreeSides side) {
+        Objects.requireNonNull(processable);
+        // Look into the cache to avoid recomputing reduced limits if they were already computed
+        // with the same limit reductions
+        CacheKey<P> cacheKey = new CacheKey<>(processable, limitType, side);
+        if (reducedLimitsCache.containsKey(cacheKey)) {
+            return Optional.of(reducedLimitsCache.get(cacheKey));
+        }
+
+        return computeLimitsWithAppliedReduction(processable, limitType, side);
+    }
+
+    /**
+     * <p>Retrieve the limits on <code>processable</code> then apply reductions on it.</p>
+     * <p>If no reductions applies on the resulting {@link LimitsContainer} must contains the same object for
+     * the original and the reduced limits.</p>
+     *
+     * @param processable the network element for which the limits must be retrieved and reduced
+     * @param limitType the type of limits to process
+     * @param side the side of the network element where to retrieve the original limits
+     * @return an object containing both the original and the reduced limits.
+     */
+    protected abstract Optional<LimitsContainer<L>> computeLimitsWithAppliedReduction(P processable, LimitType limitType, ThreeSides side);
```

Comment:
Same. if you have a good name for the interface, just a key method with a simple name is enough.

Reply:
This method was renamed as `computeUncachedLimits`, since its purpose is to compute the limits that are not yet in the cache.
This method is not meant to be used by users and is in protected access.